### PR TITLE
JS-236 Add eslint-plugin-sonarjs compatibility with ESLint 9

### PR DESCRIPTION
**Pull request by vdiez**
Fixes https://sonarsource.atlassian.net/browse/JS-236

### CODE REVIEW

File: .cirrus.yml [hunk 1]
```diff
@@ -306,8 +307,15 @@ eslint_plugin_test_task:
     - source set_maven_build_version $BUILD_NUMBER
     - cd packages/jsts/src/rules
     - npm run build
-    - cd ../../../../its/eslint-plugin-sonarjs
+    - cd ../../../../its/eslint8-plugin-sonarjs
     - npm run build
+    - npx eslint file.js || true # to check output in CI
```
Comment (ilia-kebets-sonarsource) [inline .cirrus.yml:312]: Isn't this the same that you do in the test? Is it debugging?

Comment (vdiez) [reply .cirrus.yml:312]: yes, mostly to see what is the output of running eslint, as we don't see the output of the test

File: its/eslint8-plugin-sonarjs/tseslint.config.mjs [hunk 2]
```diff
@@ -0,0 +1,8 @@
+// @ts-check
```
Comment (ilia-kebets-sonarsource) [inline its/eslint8-plugin-sonarjs/tseslint.config.mjs:1]: why use ts-check only here?

File: its/eslint9-plugin-sonarjs/file.js [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+  class C {
+    set m(a) {
+      this.a = a;
+    }
+  }
+}
+
+ */
+
+// if (something) {}
```
Comment (ilia-kebets-sonarsource) [inline its/eslint9-plugin-sonarjs/file.js:43]: Should you clean this one up? Same for the original file

File: its/eslint8-plugin-sonarjs/tseslint.config.mjs (hunk 2 above)
Comment (vdiez) [reply its/eslint8-plugin-sonarjs/tseslint.config.mjs:1]: to have type-checking in the file. thanks to that I saw a problem with missing estree typings

File: its/eslint9-plugin-sonarjs/file.js (hunk 3 above)
Comment (ilia-kebets-sonarsource) [inline its/eslint9-plugin-sonarjs/file.js:43]: should you clean this up or is it part of S125? same for original file

File: packages/jsts/src/rules/package.json [hunk 4]
```diff
@@ -35,11 +35,12 @@
     "@babel/preset-flow": "7.24.1",
     "@babel/preset-react": "7.24.1",
     "@eslint-community/regexpp": "4.10.0",
-    "@typescript-eslint/eslint-plugin": "7.7.1",
+    "@types/eslint": "8.56.10",
```
Comment (ilia-kebets-sonarsource) [inline packages/jsts/src/rules/package.json:38]: why move this from devDeps?

File: package.json [hunk 5]
```diff
@@ -83,15 +83,15 @@
     "@babel/preset-flow": "7.24.1",
     "@babel/preset-react": "7.24.1",
     "@eslint-community/regexpp": "4.10.0",
-    "@typescript-eslint/eslint-plugin": "7.7.1",
-    "@typescript-eslint/parser": "7.7.1",
-    "@typescript-eslint/utils": "7.7.1",
+    "@typescript-eslint/eslint-plugin": "7.16.1",
```
Comment (ilia-kebets-sonarsource) [inline package.json:86]: it's a bit unfortunate that we have to maintain versions of `@typescript-eslint/eslint-plugin` and `eslint-plugin-react` here and in rules/

File: packages/jsts/tests/rules/helpers/aws/cdk.test.ts [hunk 6]
```diff
@@ -60,14 +60,6 @@ import { default as cdk } from 'aws-cdk-lib';
...
       `,
     },
-    {
-      code: `
-const awsCdk = 'aws-cdk-lib';
-import cdk = require(awsCdk); // FN
-import module = cdk.aws_module;
-new module.Class();
-      `,
-    },
```
Comment (ilia-kebets-sonarsource) [inline packages/jsts/tests/rules/helpers/aws/cdk.test.ts:70]: is it the result of the TSESlint update?

File: packages/jsts/src/rules/core/index.ts [hunk 7]
```diff
@@ -22,4 +22,4 @@ import { Linter } from 'eslint';
 /**
  * ESLint core rules.
  */
-export const eslintRules = Object.fromEntries(new Linter().getRules());
+export const eslintRules = Object.fromEntries(new Linter({ configType: 'eslintrc' }).getRules());
```
Comment (ilia-kebets-sonarsource) [inline packages/jsts/src/rules/core/index.ts:25]: what does this do?

File: packages/jsts/src/rules/index.ts [hunk 8]
```diff
@@ -683,6 +683,11 @@ const recommendedConfig: FlatConfig.Config = {
     },
   },
   rules: {},
+  settings: {
+    react: {
+      version: '999.999.999',
+    },
+  },
```
Comment (ilia-kebets-sonarsource) [inline packages/jsts/src/rules/index.ts:690]: is this a default value?

File: its/eslint9-plugin-sonarjs/index.test.js [hunk 9]
```diff
@@ -0,0 +1,33 @@
+const { test } = require('node:test');
+const assert = require('node:assert');
+const spawn = require('cross-spawn');
```
Comment (ilia-kebets-sonarsource) [inline its/eslint9-plugin-sonarjs/index.test.js:3]: why use this one?

**Review by ilia-kebets-sonarsource — APPROVED** [review]
LGTM. Thank you for the extensive explanations

File: its/eslint9-plugin-sonarjs/file.js (hunk 3 above)
Comment (ilia-kebets-sonarsource) [reply its/eslint9-plugin-sonarjs/file.js:43]: yes

File: packages/jsts/src/rules/package.json (hunk 4 above)
Comment (ilia-kebets-sonarsource) [reply packages/jsts/src/rules/package.json:38]: the exported type defintions bundled with the package depend on it

File: packages/jsts/src/rules/core/index.ts (hunk 7 above)
Comment (ilia-kebets-sonarsource) [reply packages/jsts/src/rules/core/index.ts:25]: we need this when using the plugin through the Linter API

File: packages/jsts/src/rules/index.ts (hunk 8 above)
Comment (ilia-kebets-sonarsource) [reply packages/jsts/src/rules/index.ts:690]: yes

File: its/eslint9-plugin-sonarjs/index.test.js (hunk 9 above)
Comment (ilia-kebets-sonarsource) [reply its/eslint9-plugin-sonarjs/index.test.js:3]: needed to call npx from spawn in Windows

**Pull request merged**